watcher/src/watcher.rs

```diff
@@ -0,0 +1,284 @@
+        let filename = block_num_to_s3block_path(block_index)
+            .into_os_string()
+            .into_string()
+            .unwrap();
+        let url = src_url.join(&filename)?;
+
+        // Try and get the block.
+        log::debug!(
+            self.logger,
+            "Attempting to fetch block {} from {}",
+            block_index,
+            url
+        );
+        match self.transactions_fetcher.block_from_url(&url) {
+            Ok(archive_block) => {
+                log::debug!(
+                    self.logger,
+                    "Archive block retrieved for {:?} {:?}",
+                    src_url,
+                    block_index
+                );
+                if let Some(signature) = archive_block.signature {
+                    self.watcher_db.add_block_signature(
+                        src_url,
+                        block_index,
+                        signature,
+                        filename,
+                    )?;
+                } else {
+                    self.watcher_db.update_last_synced(src_url, block_index)?;
+                }
+                Ok(())
+            }
+            Err(err) => {
+                log::debug!(
+                    self.logger,
+                    "Could not sync block {} for url ({:?})",
+                    block_index,
+                    err
+                );
+                Err(WatcherError::SyncFailed)
+            }
+        }
+    }
+
+    /// Sync blocks and collect signatures.
+    ///
+    /// * `start` - starting block to sync.
+    /// * `max_blocks` - max number of blocks to sync per archive url. If None, continue polling.
+    pub fn sync_signatures(&self, start: u64, max_blocks: Option<u64>) -> Result<(), WatcherError> {
+        log::debug!(
+            self.logger,
+            "Now syncing signatures from {} to {:?}",
+            start,
+            max_blocks,
+        );
+
+        // Track whether sync failed - this catches cases where S3 is behind local ledger,
+        // which could happen if your local ledger was synced previously from different nodes
+        // than you are now watching. We track the sync failures so we can return control to the
+        // polling thread rather than continuously loop in this method.
+        let mut sync_failed: HashMap<String, bool> = self
+            .transactions_fetcher
+            .source_urls
+            .iter()
+            .map(|url| (url.as_str().to_string(), false))
+            .collect();
+
+        loop {
+            let mut last_synced: HashMap<String, u64> = self.watcher_db.last_synced_blocks()?;
+            log::debug!(self.logger, "Last synced: {:?}", last_synced);
+            log::debug!(self.logger, "max blocks: {:?}", max_blocks);
+            if let Some(max_blocks) = max_blocks {
+                last_synced.retain(|_url, block_index| *block_index < max_blocks);
```

Comment:
Shouldn't this be `start + max_blocks`?

Reply:
The start + max_blocks calculation is currently done in the sync thread, [here](https://github.com/mobilecoinofficial/mobilecoin/blob/520397707ec388e4c604c953960ff52dce20584c/watcher/src/watcher.rs#L263), so the `max_blocks` means `max_block_height`. I will update the docs and change name to disambiguate.